pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1309,3 +1332,90 @@ func initializeAnnotations(cd *hivev1.ClusterDeployment) {
 		cd.Annotations = map[string]string{}
 	}
 }
+
+// managePullSecrets() takes pull secretes in to account and returns the merged pull request
+func (r *ReconcileClusterDeployment) managePullSecrets(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (string, error) {
+
+	// For code readability let's call the pull secret in cluster deployment config as local pull secret
+	localPullSecret, _ := controllerutils.LoadSecretData(r.Client, cd.Spec.PullSecret.Name, cd.Namespace, corev1.DockerConfigJsonKey)
+
+	// Check if global pull secret from env as it comes from hive config
+	globalPullSecret := os.Getenv("GLOBAL_PULL_SECRET")
+
+	switch {
+	case globalPullSecret != "" && localPullSecret != "":
+		// Merge local pullSecret and globalPullSecret. If both pull secrets have same registry name
+		// then the merged pull secret will have registry secret from local pull secret
+		pullSecret, err := controllerutils.MergeJsons(globalPullSecret, localPullSecret, cdLog)
+		if err != nil {
+			errMsg := "unable to merge global pull secret with local pull secret"
+			cdLog.WithError(err).Error(errMsg)
+			return "", fmt.Errorf("%s : %v", errMsg, err)
+		}
+		return pullSecret, nil
+	case globalPullSecret != "":
+		return globalPullSecret, nil
+	case localPullSecret != "":
+		return localPullSecret, nil
+	default:
+		errMsg := "At least one of pull request need to be available"
+		cdLog.Error(errMsg)
+		return "", fmt.Errorf("%s", errMsg)
+	}
+}
+
+// updatePullSecretObj() adds pull secret information in cluster deployment and cluster deployment status
+func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
+	var pullSecretName string
+	var pullSecretHash, existingPullSecretHash string
+	var noExistingPullSecretObj bool
+
+	existingPullSecretObj := &corev1.Secret{}
+
+	if cd.Status.PullSecret.Name != "" {
+		pullSecretName = cd.Status.PullSecret.Name
+		err := r.Get(context.TODO(), types.NamespacedName{Name: pullSecretName, Namespace: cd.Namespace}, existingPullSecretObj)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				noExistingPullSecretObj = true
+			} else {
+				cdLog.WithError(err).Error("Error getting pull secret from cluster deployment")
+				return err
+			}
+		}
+	} else {
+		pullSecretName = apihelpers.GetResourceName(cd.Name, "pull-secret")
```

Comment:
Do we really need the secret name to be `{cluster-name}-{hash}-pull-secret`? Or can we accept `{cluster-name}-pull-secret-{hash}` as is more typically with kubernetes? If the latter is acceptable, then we should just set `GenerateName` instead of `Name` on the secret and let kube take care of things for us.

Reply:
Nevermind. I was confused about what `GetResourceName` was doing.